pandas/core/reshape/melt.py

```diff
@@ -115,7 +116,17 @@ def melt(
         # asanyarray will keep the columns as an Index
         mdata[col] = np.asanyarray(frame.columns._get_level_values(i)).repeat(N)
 
-    return frame._constructor(mdata, columns=mcolumns)
+    result = frame._constructor(mdata, columns=mcolumns)
+
+    if keep_index:
+        new_index = np.tile(frame.index, K)
```

Comment:
Index types (e.g. Categorical) and index names are not retained?

Reply:
Good observation, added.